[SPARK-20372] [ML] Word2Vec Continuous Bag of Words model

shubhamchopra opened the pull request:
## What changes were proposed in this pull request?
This adds Continuous Bag of Words implementation to Word2Vec. The implementation uses negative sampling and replicates the original implementation available [here](https://code.google.com/archive/p/word2vec/).

## How was this patch tested?
Patch tested using unit tests contributed as a part of this PR.

shubhamchopra committed:
Word2Vec CBOW + Negative Sampling implementation.

shubhamchopra committed:
Correcting the negative samples function.

shubhamchopra committed:
Correcting scala style issue.

shubhamchopra committed:
Checking to make sure neg samples is less than the vocab size.

shubhamchopra committed:
removing unused function.

shubhamchopra committed:
Adding test cases, similar to the ones for skip-gram based estimation.

AmplabJenkins commented:
Can one of the admins verify this patch?

hhbyyh commented:
Thanks for sharing the work. To help make the review easier, I would recommend:
1. Provide some background info. 
     Is the new algorithm better than the existing one and in which cases?
     compare with other lib or implementation of the algorithm.

2. Provide some description about your implementation.
     algorithm accuracy, scalability compared with the existing Word2Vec. 
     Is there any know issue or the limitation.

hhbyyh reviewed (commented):
Perhaps it's better to separate the CBOW related code in another class. 
Is it possible to leverage some work from MLP implementation?

shubhamchopra commented:
The [original paper](https://arxiv.org/abs/1301.3781) proposed two model architectures for generating word embeddings, Continuous Skip-Gram model and continuous Bag-of-words model. Spark ML currently only implements the SkipGram model. This PR adds the continuous bag of words model. As such the models compete with each other, and this implementation would give users options to settle on one which suits their data best.

The implementation is based largely on the [original C implementation](https://code.google.com/archive/p/word2vec/). I implemented this using Negative Sampling, as that was shown to have good performance [here](https://papers.nips.cc/paper/5021-distributed-representations-of-words-and-phrases-and-their-compositionality.pdf). I tried to vectorize operations using BLAS where possible.
 
I don't understand what you mean by "MLP" implementation. Can you please clarify?

holdenk commented:
Thanks for working on this, I'm traveling right now but maybe @MLNick has some bandwith to look at this.

MLnick commented:
It would be ideal to have both methods, but I'm worried about reviewer bandwidth vs priority on this.

@Krimit you were working on Word2Vec recently - thoughts? Perhaps you have time to help on review also?

Krimit commented:
I'm happy to take a look! I'll have some time to dig in deeper tomorrow. Some of my initial impressions:
* There's a lot going on here, I agree with @hhbyyh that it would be cleaner to put the CBOW code in a new class
* Can you provide some information about the practical differences between CBOW and skip-grams? Wikipedia quotes the author (I assume they mean Tomas) as saying that ``CBOW is faster while skip-gram is slower but does a better job for infrequent words``. Has this been your experience as well? How pronounced is the difference? in what cases would a user choose one over the other? I'm basically seconding @hhbyyh's comment on a more in-depth comparison experiment. The fact that the original paper has both implementations is not in itself enough of a reason for Spark to do the same, IMO

holdenk commented:
I can maybe help out a bit in a week and a bit (I've also done some poking inside of Word2Vec) but I need to wrap up some travel and Python stuff first.

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -102,6 +107,24 @@ private[feature] trait Word2VecBase extends Params
+   * Default: true (Use Skip-Gram with Hierarchical softmax)
+   * @group param
+   */
+  final val skipGram = new BooleanParam(this, "skipGram", "Use Skip-Gram model to generate word " +
```
How about making this a ``Param[String]``, ala [LDA](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala#L175)? Maybe named ``algorithm`` or something like that. That would make it clearer, and also make it easier to add any possible future algorithms (such as GloVe)

MLnick replied:
We could use `HasSolver` trait here

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -102,6 +107,24 @@ private[feature] trait Word2VecBase extends Params
+  setDefault(skipGram -> true)
+
+  /**
+   * Number of negative samples to use with CBOW based estimation
```
add something to the effect that this parameter is ignored for ``skipGram``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -127,6 +150,11 @@ final class Word2Vec @Since("1.4.0") (
+  private val batchSize = 10000
+  private val power = 0.75
+  private val maxUnigramTableSize = 100*1000*1000
+  private val unigramTableSizeFactor = 20
```
Please add some comments describing these values. Should any of them be tune-able?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+   * negative samples.
+   */
+  private def generateUnigramTable(normalizedWeights: Array[Double], tableSize: Int): Array[Int] = {
+    val table = Array.fill(tableSize)(0)
```
``new Array[Int](tableSize)``, accomplishes the same thing but faster

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+  /**
+   * Similar to InitUnigramTable in the original code. Instead of using an array of size 100 million
+   * like the original, we size it to be 20 times the vocabulary size.
```
why 20 times the vocabulary size? Also, this comment is misplaced. It belongs with the ``unigramTableSizeFactor``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+  private def generateUnigramTable(normalizedWeights: Array[Double], tableSize: Int): Array[Int] = {
+    val table = Array.fill(tableSize)(0)
+    var a = 0
+    var i = 0
```
Could you please use more descriptive variable names here? I was expecting ``i`` to be the index into the table array

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+    while (a < table.length) {
+      table.update(a, i)
+      if (a.toFloat / table.length >= normalizedWeights(i)) {
+        i = math.min(normalizedWeights.length - 1, i + 1)
```
An ``if else`` here would be much more legible

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+  }
+
+  private def generateVocab[S <: Iterable[String]](input: RDD[S]):
+      (Int, Long, Map[String, Int], Array[Int]) = {
```
This could use a named class. It's not immediately clear what each of these are

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val vocab = words.map(w => (w, 1L))
+      .reduceByKey(_ + _)
+      .filter{case (w, c) => c >= $(minCount)}
```
You should broadcast ``$(minCount)`` to avoid shipping the entire class to each task. (and make sure to destroy it at the end of the method)

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val totalWordCount = vocab.map(_._2).sum
+
+    val vocabMap = vocab.zipWithIndex.map{case ((w, c), i) =>
```
``vocab.keySet.zipWithIndex``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+      w -> i
+    }.toMap
+
+    // We create a cumulative distribution array, unlike the original implemention
```
In curious, why ``unlike the original implemention``? why not follow the original? (also, typo, should be ``implementation``)

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val words = input.flatMap(x => x)
+
+    val vocab = words.map(w => (w, 1L))
```
maybe call this ``wordCounts`` or something like that to make it clearer?

MLnick replied:
The final thing returned is the actual vocab sorted by occurrence, so `vocab` is probably better?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+  /**
+   * This method implements Word2Vec Continuous Bag Of Words based implementation using
+   * negative sampling optimization, using BLAS for vectorizing operations where applicable.
+   * The algorithm is parallelized in the same way as the skip-gram based estimation.
```
Nit: It's worthwhile to describe how this is parallelized. What happens if the skip-gram implementation changes?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+   * negative sampling optimization, using BLAS for vectorizing operations where applicable.
+   * The algorithm is parallelized in the same way as the skip-gram based estimation.
+   * @param input
+   * @return
```
nit: fill these out

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+   * @param input
+   * @return
+   */
+  private def fitCBOW[S <: Iterable[String]](input: RDD[S]): feature.Word2VecModel = {
```
I think the CBOW logic should definitely live outside of this class. I'm even wondering if it's worthwhile to update the mllib Word2Vec to have access to this as well. @MLnick @srowen thoughts?

MLnick replied:
If it's easier to add it to mllib Word2Vec that's ok. Preference is not to touch mllib if possible - because we would actually prefer to port the older mllib stuff over to live purely in ml (and to use DataFrame operations where it can potentially improve performance).

But I agree it should probably live in a separate object at the least.

MLnick replied:
There should be some level of overlap where we can re-use as much code as possible.

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+  private def fitCBOW[S <: Iterable[String]](input: RDD[S]): feature.Word2VecModel = {
+    val (vocabSize, totalWordCount, vocabMap, uniTable) = generateVocab(input)
+    val negSamples = $(negativeSamples)
+    assert(negSamples < vocabSize,
```
Is this assertion truly needed? What about just ``negSamples = min($(negSamples) - 1, vocabSize)``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+      math.min(maxUnigramTableSize, unigramTableSizeFactor * normalizedCumWeights.length)
+    val unigramTable = generateUnigramTable(normalizedCumWeights, unigramTableSize)
+
+    (vocabMap.size, totalWordCount, vocabMap, unigramTable)
```
No need to return ``vocabMap.size``, that information is duplicated from ``vocabMap``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val sc = input.context
+
+    val vocabMapbc = sc.broadcast(vocabMap)
```
nit: ``vocabMapBroadcast``, same for ``unigramTable``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val digitSentences = input.flatMap{sentence =>
+      val wordIndexes = sentence.flatMap(vocabMapbc.value.get)
+      wordIndexes.grouped($(maxSentenceLength)).map(_.toArray)
```
I think you need to broadcast ``maxSentenceLength``

MLnick replied:
If we wrap up the algo in a separate object, and feed in the `Param` values as args, we should only capture the object scope and not need to bc all the little params?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+      val partialFits = digitSentences.mapPartitionsWithIndex{ case (i_, iter) =>
+        logInfo(s"Iteration: $iteration, Partition: $i_")
+        logInfo(s"Numerical lib class being used : ${blas.getClass.getName}")
```
Are these logs really necessary?

MLnick replied:
Yeah, I think `logDebug`

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+        val gb = Array.fill(negSamples + 1)(0.0f)
+        val hiddenLayerUpdate = Array.fill(vectorSize * (negSamples + 1))(0.0f)
+        val neu1e = Array.fill(vectorSize)(0.0f)
+        val wordIndices = Array.fill(negSamples + 1)(0)
```
Looks like your initializing all of these arrays to zero. Just do ``new Array[Float/Int](size)`` which is faster

Krimit commented:
@shubhamchopra have you run this code in a distributed spark cluster yet?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+    new feature.Word2VecModel(vocabMap, syn0Global)
+  }
+
+  private def initializeVector(v: Array[Float], value: Float = 0.0f): Unit = {
```
Nit: From what I can tell ``value`` is never actually passed in, so it's always equal to ``0.0f``. Maybe call the method ``zeroVector`` and make that more explicit?

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+  private def initializeVector(v: Array[Float], value: Float = 0.0f): Unit = {
+    var i = 0
+    val length = v.length
+    while(i < length) {
```
``while (i < v.length)``

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+            blas
+              .sgemv("T", cols, rows, 1.0f, l2Vectors, 0, cols, contextVector, 0, 1, 0.0f, gb, 0, 1)
+
+            (0 to gb.length-1).foreach {i =>
```
a simple ``while`` loop would be faster

Krimit commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+
+    val window = $(windowSize)
+
+    val digitSentences = input.flatMap{sentence =>
```
I might be missing something, but what is meant by ``digitSentences``?

MLnick replied:
nit: space around `{`: `flatMap { sentence => `

shubhamchopra commented:
@Krimit @MLnick @hhbyyh I am working on getting your earlier queries answered.
@Krimit Thanks for looking into the code, I will try to get the code-review feedback incorporated in a couple of day or so.

shubhamchopra commented:
@Krimit 
_Can you provide some information about the practical differences between CBOW and skip-grams?_
![Model Architectures](https://cloud.githubusercontent.com/assets/6588487/25546610/d0f95aa8-2c31-11e7-8b47-4f9d31254f0f.png)
As mentioned in [this paper](https://arxiv.org/pdf/1301.3781.pdf), CBOW model looks at the words around a target word, and tries to predict the target word. SkipGram does just the opposite. Given a target word, it tries to predict the context words around it. The prediction is done using a very simple neural network with a single hidden layer. 

_Wikipedia quotes the author (I assume they mean Tomas) as saying that CBOW is faster while skip-gram is slower but does a better job for infrequent words. Has this been your experience as well? How pronounced is the difference?_ 
The current CBOW + Negative Sampling I found to take almost the same time as the existing SkipGram + Hierarchical sampling. The negative sampling is tunable, and the performance will be slower for a higher number of negative samples.

_in what cases would a user choose one over the other?  I'm basically seconding @hhbyyh's comment on a more in-depth comparison experiment._
There is a good amount of research around this with comparison experiments. It appears to largely depend on the application embeddings would be used for. [Levy et al](http://www.aclweb.org/anthology/Q15-1016) show how different methods perform with extensive experiments. They used the embeddings to perform similarity, relatedness and other tests on some open datasets.

[Mikolov et al](https://papers.nips.cc/paper/5021-distributed-representations-of-words-and-phrases-and-their-compositionality.pdf) found SkipGram with Negative Sampling to outperform CBOW. [Baroni et al](http://anthology.aclweb.org/P/P14/P14-1023.pdf) found that CBOW had a slight advantage. [Levy et al](http://www.aclweb.org/anthology/Q15-1016) explain that while CBOW did not perform as well in their experiments, others have shown that capturing joint contexts (CBOW does this) can improve performance on word similarity tasks. They also saw CBOW to perform well in analogy tasks. So again, it depends on the task being performed.

[Mikolov et al](https://arxiv.org/pdf/1309.4168.pdf) recommend using Skip-Gram when mono-lingual data is small and CBOW for larger datasets.

_The fact that the original paper has both implementations is not in itself enough of a reason for Spark to do the same, IMO_
This is an active area of research, and both methods generate embeddings that perform well on different tasks. As a library providing these implementations, the choice I think is best left to the user and the application it is being used for.

Krimit commented:
Thanks for the detailed response @shubhamchopra. 
I'd like to clarify my point about whether this should be implemented in Spark: Spark MlLib is first and foremost a framework for doing ML on large datasets where other existing implementations (such as ``scikit-learn``) are impractical. A reality of ML is that often increasing the size (and quality) of the training data is much more important than tweaking model hyper-parameters. Therefore as a community, I think our focus should be more on robustness than on "completeness".

While having additional algorithms available for tuning can helpful, I would personally be more interested in additions that offer significant and clear benefits (such as ``GloVe`` which should be much faster to train and a really good fit for Spark due the natural parallelization of the problem).

With that said, I'm not opposed to adding CBOW, so long as we vet it. As part of having this merged in, I think ideally we should run an experiment on a large-ish dataset (wikipedia?) comparing the two implementations

shubhamchopra committed:
incorporating code feedback from @Krimit

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+            }
+            gb.map(math.abs).sum / alpha
+          }
+          logInfo(s"Partition: $i_, Average Batch Error = ${errors.sum / batchSize}")
```
`logDebug`

MLnick commented:
FYI, realistically there won't be bandwidth to really focus on this until after Spark 2.2 QA is done at the earliest.

shubhamchopra committed:
Using HasSolver trait.

shubhamchopra committed:
Adding sub sampling and incorporating feedback.

shubhamchopra committed:
Putting the CBOW solver in a separate object.

shubhamchopra commented:
@MLnick I half expected that. No worries. I have incorporated some of your feedback in the meantime and also added subsampling as well. Thanks for looking into the code.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -36,7 +36,10 @@ import org.apache.spark.util.{Utils, VersionUtils}
+  with HasInputCol with HasOutputCol with HasMaxIter with HasStepSize with HasSeed with HasSolver {
+  // We currently support SkipGram with Hierarchical Softmax and
+  // Continuous Bag of Words with Negative Sampling
+  private val supportedModels = Array("sg-hs", "cbow-ns")
```
how is this used?

shubhamchopra replied:
Good catch. It is not used. Fixing it.

shubhamchopra committed:
Removing unused variables.

shubhamchopra committed:
Converting log statement to debug

shubhamchopra commented:
Code-review comments/suggestions so far have been incorporated. Thanks for looking into the code. Happy to incorporate more suggestions and feedback.

shubhamchopra committed:
correcting indentation.

shubhamchopra committed:
when using sampling, the sampled sentence should be used downstream.

shubhamchopra committed:
Merging master.

MLnick commented:
ok to test

SparkQA commented:
**[Test build #80243 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80243/testReport)** for PR 17673 at commit [`feda8dc`](https://github.com/apache/spark/commit/feda8dce8c2832bd1a3c61a84bfac9a23629866a).

SparkQA commented:
**[Test build #80243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80243/testReport)** for PR 17673 at commit [`feda8dc`](https://github.com/apache/spark/commit/feda8dce8c2832bd1a3c61a84bfac9a23629866a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80243/
Test PASSed.

shubhamchopra committed:
Changing the way estimated vectors are aggregated, so large vector sizes don't cause shuffle issues.

SparkQA commented:
**[Test build #81231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81231/testReport)** for PR 17673 at commit [`948cc15`](https://github.com/apache/spark/commit/948cc15b67113b8ad74b67eeef13a39f55b7313a).

SparkQA commented:
**[Test build #81231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81231/testReport)** for PR 17673 at commit [`948cc15`](https://github.com/apache/spark/commit/948cc15b67113b8ad74b67eeef13a39f55b7313a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81231/
Test FAILed.

SparkQA commented:
**[Test build #81320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81320/testReport)** for PR 17673 at commit [`361d79d`](https://github.com/apache/spark/commit/361d79ddeab78889cd5a0a63f21d1e446a7a34fd).

shubhamchopra committed:
Simplifying the inner estimation loop. This should also improve performance.

SparkQA commented:
**[Test build #81320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81320/testReport)** for PR 17673 at commit [`361d79d`](https://github.com/apache/spark/commit/361d79ddeab78889cd5a0a63f21d1e446a7a34fd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81320/
Test PASSed.

shubhamchopra committed:
Adding MAX_EXP check as in both the C implementation and the SGHS implementation.

SparkQA commented:
**[Test build #82005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82005/testReport)** for PR 17673 at commit [`64a5a6b`](https://github.com/apache/spark/commit/64a5a6b2b3cacedc82b24bde9347fee272b78849).

SparkQA commented:
**[Test build #82005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82005/testReport)** for PR 17673 at commit [`64a5a6b`](https://github.com/apache/spark/commit/64a5a6b2b3cacedc82b24bde9347fee272b78849).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82005/
Test PASSed.

MLnick reviewed (commented):
Took a high level pass and made some general comments. I still need to go through the details of the algorithm.

sethah commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -171,20 +210,46 @@ final class Word2Vec @Since("1.4.0") (
 
+  /** @group setParam */
+  @Since("2.2.0")
+  val solvers = Set("sg-hs", "cbow-ns")
```
"skipgram-hierarchical softmax"

sethah commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -171,20 +210,46 @@ final class Word2Vec @Since("1.4.0") (
 
+  /** @group setParam */
+  @Since("2.2.0")
+  val solvers = Set("sg-hs", "cbow-ns")
```
Yeah, for reference you can just look at how linear regression does the `supportedSolvers`. Also, the require isn't necessary, you can just use `ParamValidators.inArray[String](supportedSolvers))`

shubhamchopra replied:
Got it. I will have to rebase to master for this change, HasSolver trait has been updated since I implemented this. Keeping this for the end.

shubhamchopra commented:
Thanks for your comments/suggestions @MLnick and @sethah . Working on incorporating these.

sethah commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+    arr.update(0, word)
+    var i = 1
+    while (i <= numSamples) {
+      val negSample = unigramTable(random.nextInt(unigramTable.length))
```
The unigram table by default is `100e6 * 4bytes = 400 MB` in size. The original code was meant to run on a single machine, but for Spark we are going to serialize this every iteration, which may be prohibitive. AFAICT you can just simulate the behavior needed here without actually duplicating word indices over a 100million length array. Just keep the normalized cumulative weights around and here, generate a random number between 0 and 1. Then, just do a binary search to figure out which word index it is. You only need to store O(vocabSize) numbers. The time complexity increases from O(numNonNegativeSamples) to O(numNonNegativeSamples * log2(vocabSize)), but the inner loop of the training is bound by O(numNonNegativeSamples * vectorSize), which should typically be much larger. Have you tested it at scale? Does the broadcast step take a long time? Also, what are typical vocabulary sizes? A vocab size of 1e6 is going to be about 4 MB so the space savings are pretty big.

shubhamchopra replied:
I agree with the analysis. I did initially implement it using the binary search method you suggested. Changing to this reduce that complexity to constant time, and I saw 5%-10% performance gain from it. I do agree about the space savings

To answer your questions:
1. I have run it at scale, on wikipedia dump and some other data sets.
2. This broadcast doesn't take too long, and happens only once at startup. Note that the weights are also collected and broadcasted every iteration, and they are also similar in size (1e6 vocab x 100 dimensions x 4bytes ~ 400Mb). 
3. Typical vocab sizes can be 500k-1e6. As a comparison [GloVe Vectors](https://nlp.stanford.edu/projects/glove/) vocab sizes range from 400k for wikipedia to 2e6 for web-crawl and such.

sethah reviewed (commented):
First pass. Didn't make it to the tests. Thanks for contributing this!

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -106,6 +106,45 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 15
+   * @group param
+   */
+  final val negativeSamples = new IntParam(this, "negativeSamples", "Number of negative samples " +
```
`negativeSamples` is only used when Negative Sampling is used with CBOW (or, as in #18123, SkipGram)
SkipGram with Hierarchical Softmax, which was the initial version, is not affected by this parameter.

In principle, we now have 3 possible techniques that can be used (another one implemented in #18123). Open to suggestions possible values for `solver`

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -106,6 +106,45 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 15
+   * @group param
+   */
+  final val negativeSamples = new IntParam(this, "negativeSamples", "Number of negative samples " +
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -106,6 +106,45 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 0.0
+   * @group param
+   */
+  final val sample = new DoubleParam(this, "samplingThreshold", "Sampling threshold to reduce " +
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -106,6 +106,45 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 0.0
+   * @group param
+   */
+  final val sample = new DoubleParam(this, "samplingThreshold", "Sampling threshold to reduce " +
```
Corrected. Thanks!

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -171,20 +210,46 @@ final class Word2Vec @Since("1.4.0") (
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setMaxUnigramTableSize(value: Int): this.type = set(unigramTableSize, value)
```
Done

shubhamchopra commented on mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala:
```diff
@@ -189,6 +305,136 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext with Defaul
+      case Row(w: String, sim: Double) => (w, sim)
+    }.collect().unzip
+
+    // Increase the window size
```
Corrected. Thanks.

shubhamchopra commented on mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala:
```diff
@@ -245,5 +508,28 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext with Defaul
     model.transform(ngramDF).collect()
   }
 
+  test("Word2Vec works with input that is non-nullable (NGram) - CBOW") {
```
Yes. Taking another pass on the tests to remove code duplication.

shubhamchopra committed:
Merging master

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+import org.apache.spark.rdd.RDD
+import org.apache.spark.util.random.XORShiftRandom
+
+object Word2VecCBOWSolver extends Logging {
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+      word2Vec: Word2Vec,
+      input: RDD[S]): feature.Word2VecModel = {
+
+    val negativeSamples = word2Vec.getNegativeSamples
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+      input: RDD[S]): feature.Word2VecModel = {
+
+    val negativeSamples = word2Vec.getNegativeSamples
+    val sample = word2Vec.getSample
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+
+    val Vocabulary(totalWordCount, vocabMap, uniTable, sampleTable) =
+      generateVocab(input, word2Vec.getMinCount, sample, word2Vec.getUnigramTableSize)
+    val vocabSize = vocabMap.size
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+    val initRandom = new XORShiftRandom(seed)
+
+    val vectorSize = word2Vec.getVectorSize
+    val syn0Global = Array.fill(vocabSize * vectorSize)(initRandom.nextFloat - 0.5f)
```
There are a few places where the implementation seems counter-intuitive, and doesn't directly relate to the papers its based on. I thought it might help if someone wants to cross-check the implementation with the C implementation. Other than that, I do agree with you, and have tried to name things appropriately, other than these key variables around which the implementation is based.

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+    logInfo(s"VocabSize: ${vocabMap.size}, TotalWordCount: $totalWordCount")
+
+    val maxIter = word2Vec.getMaxIter
+    for {iteration <- 1 to maxIter} {
```
It's not needed here, is it?

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+
+        val groupedBatches = contextWordPairs.grouped(batchSize)
+
+        val negLabels = 1.0f +: Array.fill(negativeSamples)(0.0f)
```
I instantiate them once and reuse them for the entire iteration. This gave significant speed-ups.

sethah replied:
instead of `val err = (negLabels(i) - v) * alpha`, using `val err = ((if (i > 0) 0f else 1f) - v) * alpha`. Then you don't need to allocate the array?

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+      val syn0bc = sc.broadcast(syn0Global)
+      val syn1bc = sc.broadcast(syn1Global)
+
+      val partialFits = digitSentences.mapPartitionsWithIndex { case (i_, iter) =>
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+
+        // initialize intermediate arrays
+        val contextVec = new Array[Float](vectorSize)
+        val l2Vectors = new Array[Float](vectorSize * (negativeSamples + 1))
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+        val l2Vectors = new Array[Float](vectorSize * (negativeSamples + 1))
+        val gb = new Array[Float](negativeSamples + 1)
+        val neu1e = new Array[Float](vectorSize)
+        val wordIndices = new Array[Int](negativeSamples + 1)
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+              (maxIter * (wordsPerPartition.toFloat + 1)) + ((iteration - 1).toFloat / maxIter)
+          val alpha = math.max(learningRate * 0.0001, learningRate * (1 - wordRatio)).toFloat
+
+          if(idx % 10 == 0 && idx > 0) {
```
Good catch. Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+        val time = System.nanoTime
+        var batchTime = System.nanoTime
+        var idx = -1L
+        for (batch <- groupedBatches) {
```
Done

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,344 @@
+
+          val errors = for ((contextIds, word) <- batch) yield {
+            // initialize vectors to 0
+            zeroVector(contextVec)
```
Done

SparkQA commented:
**[Test build #82568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82568/testReport)** for PR 17673 at commit [`236e4c1`](https://github.com/apache/spark/commit/236e4c1db2051f2f8a0435e753df3579afdfeb5e).

SparkQA commented:
**[Test build #82568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82568/testReport)** for PR 17673 at commit [`236e4c1`](https://github.com/apache/spark/commit/236e4c1db2051f2f8a0435e753df3579afdfeb5e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82568/
Test FAILed.

shubhamchopra committed:
Incorporating suggestions/feedback from @MLnick and @sethah

SparkQA commented:
**[Test build #82569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82569/testReport)** for PR 17673 at commit [`9090b96`](https://github.com/apache/spark/commit/9090b967e03e43e3a709d9c2c94fe75de5b9a8e6).

shubhamchopra committed:
Correcting style errors.

SparkQA commented:
**[Test build #82569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82569/testReport)** for PR 17673 at commit [`9090b96`](https://github.com/apache/spark/commit/9090b967e03e43e3a709d9c2c94fe75de5b9a8e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82569/
Test PASSed.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+
+  /**
+   * Solver used for Word2Vec.
+   * Supported options are "sg-hs" and "cbow-ns"
```
add more details:
sg-hs: Skip-gram with hierarchical softmax
cbow: ...

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
 
+  /**
+   * Number of negative samples to use with CBOW based estimation.
+   * This parameter is ignored for SkipGram-Hierachical Softmax based estimation.
```
Rather than exclusion, I would recommend to use precise description here and in other places.
The parameter is only used for continuous bag-of-words with negative sampling algorithm, i.e. solver = "cbow-ns".

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 100 million
+   * @group param
+   */
+  final val unigramTableSize = new IntParam(this, "unigramTableSize", "Max table size to " +
```
expert param?

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+   */
+  final val unigramTableSize = new IntParam(this, "unigramTableSize", "Max table size to " +
+    "use for generating negative samples", ParamValidators.gt(1))
+  setDefault(unigramTableSize -> 100*1000*1000)
```
maybe move the setDefault after setter in Word2Vec, which seems to be a recommended practice right now. (Just as the solver)

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+   * Default: 0.0
+   * @group param
+   */
+  final val samplingThreshold = new DoubleParam(this, "samplingThreshold", "Sampling threshold" +
```
perhaps we can rename this to a more precise name, E.g. downSamplingTFThreshold or highTFThreshold

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+   */
+  final val samplingThreshold = new DoubleParam(this, "samplingThreshold", "Sampling threshold" +
+    " to reduce words with high frequencies", ParamValidators.gtEq(0.0))
+  setDefault(samplingThreshold -> 0.0)
```
IMO we can use 1e-5 here. Default param should target for best algorithm outcome.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/impl/Word2VecCBOWSolver.scala:
```diff
@@ -0,0 +1,371 @@
+
+private [feature] object Word2VecCBOWSolver extends Logging {
+  // learning rate is updated for every batch of size batchSize
+  private val batchSize = 10000
```
IMO we should expose batchSize.

hhbyyh reviewed (commented):
Made an initial pass on API. Thanks for the PR. 

For the implementation part, since we're already using batch, is it possible that we migrate it to Level 3 BLAS operation, which could be orders of magnitudes faster.

And by MLP I meant MultilayerPerceptron.

shubhamchopra commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -105,6 +106,56 @@ private[feature] trait Word2VecBase extends Params
+   */
+  final val numNegativeSamples = new IntParam(this, "numNegativeSamples", "Number of negative" +
+    " samples to use with CBOW estimation", ParamValidators.gt(0))
+  setDefault(numNegativeSamples -> 15)
```
15 gives better performance than 5. I agree with your point below as well. We should make the default parameters ones that result in best performance.

shubhamchopra commented:
@hhbyyh Thanks for your suggestions. Will try to incorporate these in a day or so.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins OK to test.

holdenk commented:
@shubhamchopra  are you still working on this?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

ngopal commented:
When can we anticipate this branch being merged?

srowen commented:
@ngopal this one can't be merged as-is and looks like it was abandoned. Would you like to take this PR, update per reviews? I'd review that. I think CBOW could be useful in MLlib.
